iceoryx2-bb/lock-free/src/spmc/unrestricted_atomic.rs

```diff
@@ -73,26 +74,142 @@ impl<T: Copy> Producer<'_, T> {
 
 unsafe impl<T: Copy> Send for Producer<'_, T> {}
 unsafe impl<T: Copy> Sync for Producer<'_, T> {}
 
+#[doc(hidden)]
+#[repr(C)]
+pub struct UnrestrictedAtomicMgmt {
+    write_cell: IoxAtomicU32,
+    has_producer: IoxAtomicBool,
+}
+
+impl Default for UnrestrictedAtomicMgmt {
+    fn default() -> Self {
+        Self {
+            write_cell: IoxAtomicU32::new(1),
+            has_producer: IoxAtomicBool::new(true),
+        }
+    }
+}
+
+impl UnrestrictedAtomicMgmt {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// # Safety
+    ///
+    ///   * store operations are only allowed when this method returns Ok
+    ///   * __internal_release_producer must be called when the UnrestrictedAtomicMgmt (used
+    ///     without UnrestrictedAtomic) is dropped
+    pub unsafe fn __internal_acquire_producer(&self) -> Result<bool, bool> {
+        self.has_producer
+            .compare_exchange(true, false, Ordering::Relaxed, Ordering::Relaxed)
+    }
+
+    /// # Safety
+    ///
+    ///   * store operations are not allowed after this method was called
+    ///   * __internal_acquire_producer must have been successfully called before
+    pub unsafe fn __internal_release_producer(&self) {
+        self.has_producer.store(true, Ordering::Relaxed);
+    }
+
+    /// # Safety
+    ///
+    ///   * __internal_acquire_producer must have been successfully called before
+    ///   * the memory position must not be modified after __internal_update_write_cell has been
+    ///     called
+    pub unsafe fn __internal_get_ptr_to_write_cell(
+        &self,
+        value_size: usize,
+        value_alignment: usize,
+        data_ptr: *mut u8,
+    ) -> *mut u8 {
+        let write_cell = self.write_cell.load(Ordering::Relaxed);
+        let data_cell_ptr = align(
+            unsafe { data_ptr.add(value_size * (write_cell as usize % NUMBER_OF_CELLS)) } as usize,
+            value_alignment,
+        );
+        data_cell_ptr as *mut u8
+    }
+
+    /// # Safety
+    ///
+    ///   * the method must not be called without first writing to the memory position returned by
+    ///     __internal_get_ptr_to_write_cell
+    pub unsafe fn __internal_update_write_cell(&self) {
+        /////////////////////////
+        // SYNC POINT - write
+        // After writing the content of the write_cell, the content needs to be synced with the
+        // reader.
+        /////////////////////////
+        self.write_cell.fetch_add(1, Ordering::Release);
+    }
+
+    /// # Safety
+    ///
+    ///   * see Safety section of core::ptr::copy_nonoverlapping
+    pub unsafe fn load(
+        &self,
+        value_ptr: *mut u8,
+        value_size: usize,
+        value_alignment: usize,
+        data_ptr: *const u8,
+    ) {
+        /////////////////////////
+        // SYNC POINT - read
+        /////////////////////////
+        let mut read_cell = self.write_cell.load(Ordering::Acquire) - 1;
+
+        loop {
+            let data_cell_ptr = align(
```

Comment:
This here and line 141 is the same complex pointer operation. We should move this into a free function and test that function additionally.